### DISCUSSION TITLE
Losing Routing Points - Part II.

### DISCUSSION THREAD

**Discussion by rsoika** [eclipse-glsp/glsp#1435] [Q&A] [answered]
In my [last question](https://github.com/eclipse-glsp/glsp/discussions/1433) I described a routing behavior losing waypoints. As I described it was more a problem with the `ChangeRoutingPointsOperation` and the `ComputedBoundsAction` and the fact that in BPMN the anchor points need to be part of the 'waypoint' list which is not relevant to GLSP/Sprotty routing.

Now I try again to describe the core problem one more:

When you have a edge with custom routing points like in the picture below, GLSP provides only one routing point. In BPMN we store the model with 3 routing points as we include the anchor points too.
When I move the element, the router ignores the anchor point which leads to a bed effect:

![Peek 2024-12-04 06-32](https://github.com/user-attachments/assets/0a73ee41-f839-4f0b-98e3-9b65da5e69eb)

It is difficult for me to describe the problem clearly. But I think the core of my issue are the missing anchor points. So it may be not a problem that can be solved on the server part. It seems to be more inside the router on the client side. 

My expectation would be that the anchor point is protected by the router during moving the element. But as you can see the anchor point (that does not exist in my GModel) is lost during moving. 

Now I wonder if this is a general problem that can't be easily solved in GLSP (only be rewriting the router) or is it just the fact that I need to (somehow) define the anchor points also in the GModel ?

**Comment by rsoika** [comment]
I did more testing and also tried a solution with custom [Anchor Points](https://www.open-bpmn.org/glsp-client/SPROTTY_ANCHORS_AND_PORTS.html). Finally I came to the conclusion, that the problem is routing algorithm of the Manhattan router itself.

So my final question here is: How can I extend the GLSP Manhattan router or implement my own routing algorithm ?

**Comment by tortmayr** [comment, marked as answer]
Hi @rsoika,

to customize the `GLSPManhattanEdgeRouter` simply create a subclass that extend `GLSPManhattanEdgeRouter` and rebind it accordingly in your diagram module
```ts
rebind(GLSPManhattanEdgeRouter).to(YoucCustomRouter).inSingletonScope();
```

Alternatively if you want to create our own routing algorithm you can configure a new `IEdgeRouter`. There is an abstract base class (`AbstractEdgeRouter`) that you can use as starting point.
Note that you also of  to configure `IAnchorComputer` for your custom router. Typically you need at least anchor computer for the Ellipse,Rectangle and Diamond anchors:

```ts
@injectable()
export class MyCustomRouter extends AbstractEdgeRouter {
    static readonly KIND = 'custom';

 //...
}

@injectable()
export class CustomRectangleAnchorComputer implements IAnchorComputor {

    static KIND = MyCustomRouter.KIND + ':' + RECTANGULAR_ANCHOR_KIND;
    //...
 }
 
 /...

// di config
bind(TYPES.IEdgeRouter).to(MyCustomRouter);
bind(TYPES.IAnchorComputer).to(CustomRectangleAnchorComputer)
bind(TYPES.IAnchorComputer).to(CustomEllipticAnchorComputer)
bind(TYPES.IAnchorComputer).to(CustomDiamonAnchorComputer)

```
